src/tarantool-gdb.py

```diff
@@ -1197,4 +1206,154 @@ def invoke(self, arg, from_tty):
             s = mp.to_string()
         gdb.write(s + '\n')
 
+
+class JsonTokenPrinter:
+    '''Print a json_token object.'''
+
+    JSON_TOKEN_NUM = gdb.parse_and_eval('JSON_TOKEN_NUM')
+    JSON_TOKEN_STR = gdb.parse_and_eval('JSON_TOKEN_STR')
+    JSON_TOKEN_ANY = gdb.parse_and_eval('JSON_TOKEN_ANY')
+    JSON_TOKEN_END = gdb.parse_and_eval('JSON_TOKEN_END')
+
+    tuple_field_type = gdb.lookup_type('tuple_field')
+
+    def __init__(self, val):
+        self.val = val
+
+    def to_string(self):
+        type = self.val['type']
+        hash = self.val['hash']
+
+        if type == self.JSON_TOKEN_NUM:
+            s_id = ' id=' + str(self.val['num'])
+        elif type == self.JSON_TOKEN_STR:
+            s_id = ' id=' + self.val['str'].string('utf-8', 'strict', int(self.val['len']))
+        else:
+            s_id = str()
+        s = 'type={}{} hash={} parent={} sibling_idx={}'.format(
+            int(type),
+            s_id,
+            int(hash),
+            str(self.val['parent']),
+            int(self.val['sibling_idx']),
+        )
+
+        if self.val['parent'] != 0:
+            field = container_of(self.val.address, self.tuple_field_type, 'token')
+            s += ' field_id={} field_type={} offset_slot={} is_key_part={}/{}'.format(
+                str(field['id']),
+                str(field['type']),
+                str(field['offset_slot']),
+                str(field['is_key_part']),
+                str(field['is_multikey_part']),
+            )
+
+        return s
+
+    def children(self):
+        for i in range(0, int(self.val['max_child_idx']) + 1):
+            yield str(i), self.val['children'][i].dereference()
+
+    def display_hint(self):
+        return 'array'
+
+pp.add_printer('JsonToken', '^json_token$', JsonTokenPrinter)
+
+
+class TuplePrinter:
+    '''Print a tuple object.'''
+
+    tuple_type = gdb.lookup_type('struct tuple')
+
+    tuple_formats_sym = gdb.lookup_global_symbol('tuple_formats')
+    if not tuple_formats_sym:
+        raise NameError('tuple_formats is missing')
+    tuple_formats = tuple_formats_sym.value()
+
+    ptr_char = gdb.lookup_type('char').pointer()
+    ptr_uint32 = gdb.lookup_type('uint32_t').pointer()
+
+    # Printer configuration.
+    mp_depth = -1
+    mp_maxlen = -1
+
+    def __init__(self, val):
+        if val.type != self.tuple_type:
+            raise gdb.GdbError('expression doesn\'t evaluate to tuple')
+        self.val = val
+
+    def is_compact(self): # tuple_is_compact
+        return self.val['data_offset_bsize_raw'] & 0x8000
```

Comment:
This field exists since 2.9.0-262-g74177dd8a (74177dd8a98c6adbad1386026fdf47256722ee4e). There were other `struct tuple` layout changes, I remember at least 2.11.0-entrypoint-166-g9da70207a (9da70207aa5c8a2e8dca78ce34a425b18e4711bc) and 2.9.0-259-g452692116 (45269211658dbfff6120e16f2425e31cb0cb577d). That's likely not all.

I hope that the same script should be suitable to work with different tarantool versions, otherwise it would be a bit inconvenient: each time we would need to checkout specific script version.

Of course, it will complicate the code... But it seems valuable.

Reply:
Totally agree that different versions of tarantool should be supported. For now the script only deals with the absence of some MsgPack extensions, but I'll check the changes you mention and will adjust the code. I'll do my best to discover similar issues as well but if I miss something it's not a problem to fix later.

Reply:
done